proposals/params-span.md

```diff
@@ -21,125 +22,156 @@ public static class Console
+        log.Log("{0}, {1}", k, v);
 
-Alternative approaches may require runtime support.
+    log.Log("Count = {0}", dictionary.Count);
+}
+```
 
-### Explicit `stackalloc`
-Should we allow explicit stack allocation of arrays of managed types with `stackalloc` as well?
+The method could be lowered to:
 ```csharp
-public static ImmutableArray<TResult> Select<TSource, TResult>(this ImmutableArray<TSource> source, Func<TSource, TResult> map)
+static void LogDictionary<K, V>(this Logger log, Dictionary<K, V> dictionary)
 {
-    int n = source.Length;
-    Span<TResult> result = n <= 16 ? stackalloc TResult[n] : new TResult[n];
-    for (int i = 0; i < n; i++)
-        result[i] = map(source[i]);
-    return ImmutableArray.Create(result); // requires ImmutableArray.Create<T>([DoesNotEscape] ReadOnlySpan<T> items)
+    FixedSizeArray2<object> _tmp = new FixedSizeArray2<object>();
+
+    log.Log("Dictionary",
+        new ReadOnlySpan<object>(Array.Empty<object>()); // no reuse
+
+    foreach (var (k, v) in dictionary)
+    {
+        _tmp.Array[0] = k;
+        _tmp.Array[1] = v;
+        log.Log("{0}, {1}",
+            new ReadOnlySpan(_tmp.Array)); // reuse
+        Array.Clear(_tmp.Array);           // clear
+    }
+
+    _tmp.Array[0] = dictionary.Count;
+    log.Log("Count = {0}",
+        new ReadOnlySpan(_tmp.Array, 0, 1)); // reuse slice
 }
 ```
 
-This would require runtime support for stack allocation of arrays of non-constant length and any type, and GC tracking of the elements.
+## Open issues
+### Support `params scoped T[]`?
+Allow a `params T[]` to be marked as `scoped` and allocate argument arrays on the stack at call sites? That would avoid heap allocation at each call site, but allocations could only be reused at call sites with matching argument type _and length_.
 
-Direct runtime support for stack allocation of arrays of managed types might be useful for lowering implicit allocation as well.
+### Support `params Span<T>`?
+Support `params Span<T>` to allow the `params` method to modify the span contents, even though the effects are only observable at call sites that explicitly allocate the argument array?
 
-The GC does not currently track the lifetime of a `stackalloc` array so if the contents of the array have a shorter lifetime than the method, the compiler will need to zero the contents of the array so the lifetime of elements matches expectations.
+### Support `params IEnumerable<T>`, etc.?
+If we're extending `params` to support `ReadOnlySpan<T>`, should we also support `params` parameters of other collection types, including interfaces and concrete types? The reason to support `params ReadOnlySpan<T>` is to improve the performance of existing callers. And other collection types are already well supported by having non-`params` overloads for the other types in addition to a `params T[]` overload. That said, this proposal doesn't prevent extending `params` to other types in the future.
```

Comment:
We should mention that we think collection literals reduces the motivation here. We basically want params ROS in order to make existing call sites behave better in more cases. But in general and for new code we think it's good for people to just wrap the params elements in `[]`.